wallet/txauthor/author.go

```diff
@@ -0,0 +1,200 @@
+type AuthoredTx struct {
+	Tx          *wire.MsgTx
+	PrevScripts [][]byte
+	TotalInput  btcutil.Amount
+	ChangeIndex int // negative if no change
+}
+
+// ChangeSource provides P2PKH change output scripts for transaction creation.
+type ChangeSource func() ([]byte, error)
+
+// NewUnsignedTransaction creates an unsigned transaction paying to one or more
+// non-change outputs.  An appropiate transaction fee is included based on the
+// transaction size.
+//
+// Transaction inputs are chosen from repeated calls to fetchInputs with
+// increasing targets amounts.
+//
+// If any remaining output value can be returned to the wallet via a change
+// output without violating mempool dust rules, a P2PKH change output is
+// appended to the transaction outputs.  Since the change output may not be
+// necessary, fetchChange is called zero or one times to generate this script.
+// This function must return a P2PKH script or smaller, otherwise fee estimation
+// will be incorrect.
+//
+// If successful, the transaction, total input value spent, and all previous
+// output scripts are returned.  If the input source was unable to provide
+// enough input value to pay for every output any any necessary fees, an
+// InputSourceError is returned.
+//
+// BUGS: Fee estimation may be off when redeeming non-compressed P2PKH outputs.
+func NewUnsignedTransaction(outputs []*wire.TxOut, relayFeePerKb btcutil.Amount,
+	fetchInputs InputSource, fetchChange ChangeSource) (*AuthoredTx, error) {
+
+	targetAmount := h.SumOutputValues(outputs)
+	estimatedSize := txsizes.EstimateSerializeSize(1, outputs, true)
+	targetFee := txrules.FeeForSerializeSize(relayFeePerKb, estimatedSize)
+
+	for {
+		inputAmount, inputs, scripts, err := fetchInputs(targetAmount + targetFee)
+		if err != nil {
+			return nil, err
+		}
+		if inputAmount < targetAmount+targetFee {
+			return nil, insufficientFundsError{}
+		}
+
+		maxSignedSize := txsizes.EstimateSerializeSize(len(inputs), outputs, true)
+		maxRequiredFee := txrules.FeeForSerializeSize(relayFeePerKb, maxSignedSize)
+		remainingAmount := inputAmount - targetAmount
+		if remainingAmount < maxRequiredFee {
+			targetFee = maxRequiredFee
+			continue
+		}
+
+		unsignedTransaction := &wire.MsgTx{
+			Version:  wire.TxVersion,
+			TxIn:     inputs,
+			TxOut:    outputs,
+			LockTime: 0,
+		}
+		changeIndex := -1
+		changeAmount := inputAmount - targetAmount - maxRequiredFee
+		if !txrules.IsDust(changeAmount, txsizes.P2PKHPkScriptSize, relayFeePerKb) {
+			changeScript, err := fetchChange()
+			if err != nil {
+				return nil, err
+			}
+			if len(changeScript) > txsizes.P2PKHPkScriptSize {
+				return nil, errors.New("fee estimation requires change " +
+					"scripts no larger than P2PKH output scripts")
+			}
+			change := wire.NewTxOut(int64(changeAmount), changeScript)
+			l := len(outputs)
+			unsignedTransaction.TxOut = append(outputs[:l:l], change)
+			changeIndex = l
+		}
+
+		return &AuthoredTx{
+			unsignedTransaction,
```

Comment:
It also hides bugs if you miss a field.  I don't agree with the readability argument either (none of the types are the same, so it's impossible to assign them to the wrong field, and you can be assured that none of the fields have been missed).

But I'll change it because others feel strongly about it and I'm not about to argue a non issue.